uwrt_utils/include/uwrt_utils/uwrt_can.h

```diff
@@ -0,0 +1,135 @@
+#include <map>
+#include <mutex>
+#include <stdexcept>
+#include <thread>
+#include <vector>
+
+namespace uwrt_utils {
+
+class UWRTCANWrapper {
+ public:
+  explicit UWRTCANWrapper() = default;
+  explicit UWRTCANWrapper(std::string name, std::string interface_name, bool rcv_big_endian,
+                          int thread_sleep_millis = 1);
+  explicit UWRTCANWrapper(const UWRTCANWrapper& to_copy) = delete;
+  // NOLINTNEXTLINE(performance-noexcept-move-constructor, bugprone-exception-escape)
+  explicit UWRTCANWrapper(UWRTCANWrapper&& to_move);
+
+  ~UWRTCANWrapper();
+
+  UWRTCANWrapper& operator=(const UWRTCANWrapper& to_copy) = delete;
+  // NOLINTNEXTLINE(performance-noexcept-move-constructor, bugprone-exception-escape)
+  UWRTCANWrapper& operator=(UWRTCANWrapper&& to_move);
+
+  void init(const std::vector<uint32_t>& ids);
+
+ private:
+  // description name for this wrapper
+  std::string name_;
+
+  // variables needed for can socket
+  std::string interface_name_;
+  int socket_handle_{};
+  struct sockaddr_can sock_addr_ {};
+  struct ifreq ifr_ {};
+
+  // variables for wrapper
+  bool initialized_{false};
+  int rcv_endianness_{__BYTE_ORDER__};
+  std::thread read_thread_;
+  volatile bool read_thread_running_{};
+  std::map<canid_t, struct can_frame> recv_map_;
+  std::timed_mutex recv_map_mtx_;
+  std::chrono::milliseconds thread_sleep_millis_{};
+  static constexpr std::chrono::milliseconds MUTEX_LOCK_TIMEOUT{1};
+
+  // read function to be run in the thread
+  void readSocketTask();
+
+  // function to get swap endianness
+  template <class T>
+  T correctEndianness(const T& data) {
+    // if endianness is the same, just return the data back
+    if (rcv_endianness_ == __BYTE_ORDER__) {
+      return data;
+    }
+
+    // if not, we need to swap endianess
+    T swp_data;
+    auto swp_ptr = reinterpret_cast<uint8_t*>(&swp_data);      // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+    auto unswp_ptr = reinterpret_cast<const uint8_t*>(&data);  // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+    for (int i = 0; i < sizeof(T); i++) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      std::memcpy(&swp_ptr[i], &unswp_ptr[sizeof(T) - 1 - i], 1);
+    }
+    return swp_data;
+  }
+
+ public:
+  // these two function definitions need to be in header because they use templates (good one c++ 11)
+  template <class T>
+  bool getLatestFromID(T& data, uint32_t id) {
+    // make sure we have been initialized
+    if (!initialized_) {
+      throw std::runtime_error("Please initialize CAN wrapper before using it");
+    }
+
+    // check that we have new data to read at specified id
+    std::map<canid_t, struct can_frame>::iterator it;
+    if (!recv_map_mtx_.try_lock_for(MUTEX_LOCK_TIMEOUT)) {
+      throw std::runtime_error("Timedout while trying to lock read mutex");
+    }
+    it = recv_map_.find((canid_t)id);
+    if (it == recv_map_.end()) {
+      recv_map_mtx_.unlock();
```

Comment:
locks of unlocks in this function. while the logic is 100% sound rn, it may be safer to use a RAII style lock. take a look at this [SO answer](https://stackoverflow.com/a/25033678)

Reply:
Added issue #63